# 02 - Python basics B

## 字符串Strings

### 创建

In [2]:
# single quote/double是一样的
"string" == 'string'

True

In [1]:
"string'", 'string"'

("string'", 'string"')

In [11]:
"""string'""", '''string"'''

("string'", 'string"')

In [3]:
len("next")

4

In [4]:
len("\next")

4

**转义字符(escape character)**: 反斜杠 `\` 开头的特殊字符


| 转义字符 | 说明                          | 示例                | 输出结果          |
|----------|-----------------------------|---------------------|------------------|
| `\\`     | 反斜杠                      | `"C:\\Users"`       | `C:\Users`       |
| `\'`     | 单引号                      | `'It\'s me'`       | `It's me`        |
| `\"`     | 双引号                      | `"She said \"Hi\""` | `She said "Hi"`  |
| `\n`     | 换行符                      | `"Hello\nWorld"`   | `Hello`<br>`World` |
| `\t`     | 水平制表符 (Tab)            | `"Name:\tAlice"`   | `Name:    Alice` |


In [23]:
"She said \"Hi\""

'She said "Hi"'

In [7]:
print("She said \"Hi\"")

She said "Hi"


In [25]:
a = "C:\\data"
a  # 以代码的形式输出

'C:\\data'

In [22]:
print(a)

C:\data


In [29]:
# 三重引号创建的字符串支持换行
x = """
This is a string.
"""
x

'\nThis is a string.\n'

**原始字符串(raw-string)** 是以字母`r`开头的特殊字符串，转义机制在这里不起作用。

In [8]:
print(r'She said \"Hi\"')

She said \"Hi\"


In [12]:
len(r'\next')

5

### 基本操作

In [30]:
'abc' + 'def'

'abcdef'

In [31]:
'abc' 'def' 'ghi'

'abcdefghi'

In [39]:
name = "LI Lei"
print("Hello, " + name + "!")

Hello, LI Lei!


In [40]:
print(f"Hello, {name}!")

Hello, LI Lei!


In [32]:
'abc' * 2

'abcabc'

In [78]:
list('abc')

['a', 'b', 'c']

In [79]:
list('abc') * 2

['a', 'b', 'c', 'a', 'b', 'c']

In [36]:
'abc'[::-1]

'cba'

In [38]:
'a' < 'b', 'a' == 'a', 'a ' > 'a'

(True, True, True)

补充：方法（method）即绑定某个对象的函数
- `object.attribute`
- `function(arguments)`

`object.method(argument)`

In [44]:
# 列出x对象的所有attribute
print(dir(x))

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'removeprefix', 'removesuffix', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


In [41]:
x = 'abc-abc-ab'
x.replace('ab', 'xy')

'xyc-xyc-xy'

In [77]:
x.replace('ab', 'xy', 1)

'xyc-abc-ab'

In [45]:
x.capitalize()

'Abc-abc-ab'

In [47]:
x.find('a')

0

In [48]:
'123'.isdecimal()

True

In [49]:
'123'.isdigit()

True

In [53]:
'    '.isspace()

True

In [57]:
','.join(['123', '456', '789'])

'123,456,789'

In [58]:
'123,456,789'.split(',')

['123', '456', '789']

In [62]:
'123,456,789'.split(',', 1)

['123', '456,789']

In [60]:
'123,456,789'.rsplit(',', 1)

['123,456', '789']

In [63]:
'abc'.startswith('ab'), 'abc'.endswith('x')

(True, False)

In [64]:
'abc'.upper(), 'Abc'.lower()

('ABC', 'abc')

In [74]:
'aabbccabc'.translate(str.maketrans('abc', 'xyz'))

'xxyyzzxyz'

In [83]:
x = 'abc\txyz\t3.14\trst\n'
print(x)

abc	xyz	3.14	rst



In [84]:
x.split()

['abc', 'xyz', '3.14', 'rst']

#### 字符串格式化

(已经讲过f-string)

In [102]:
"Hello, {} and {}!".format(name, "HAN Meimei")

'Hello, LI Lei and HAN Meimei!'

In [103]:
"pi = %05.2f, r = %06.02f" % (3.1, 2.1324)

'pi = 03.10, r = 002.13'

In [104]:
# "02"表示小数点后面精确到2位小数，不足在后面补0
# "05"中的5表示总的字符串宽度，0表示不足在前面补0
len("%05.2f" % (3.1))

5

#### CSV与TSV的文本解析

In [87]:
x = 'abc,xyz,3.14,rst\n'
x.split(',')

['abc', 'xyz', '3.14', 'rst\n']

In [88]:
x.rstrip()

'abc,xyz,3.14,rst'

In [90]:
x.rstrip().split(',')

['abc', 'xyz', '3.14', 'rst']

### 正则表达式 (regular expression)

python对正则表达式的支持：
- https://docs.python.org/3/library/re.html
- https://docs.python.org/3/howto/regex.html#regex-howto

怎么写正则表达式：LLM可能比大多数程序员效率高。

🙋**练习：自学python正则表达式**
- https://realpython.com/regex-python
- https://realpython.com/regex-python-part-2/
- https://www.runoob.com/python3/python3-reg-expressions.html （中文）

## 函数(function)也是对象

```python
def function_name(arg1, arg2, ...):
    ... 
    return 
```

In [155]:
def function_name(arg1, arg2):
    """
    documentation
    """
    result = [arg1, arg2]
    return result

## 对象的特性(attributes)

```python
object.attribute
```
python中的对象有很多特性，特性可以是普通的对象，即属性(property)，也可以是适用于该对象的函数,即方法（method）。

In [31]:
# dir函数可以查看函数的所有特性
help(dir)

Help on built-in function dir in module builtins:

dir(...)
    dir([object]) -> list of strings

    If called without an argument, return the names in the current scope.
    Else, return an alphabetized list of names comprising (some of) the attributes
    of the given object, and of attributes reachable from it.
    If the object supplies a method named __dir__, it will be used; otherwise
    the default dir() logic is used and returns:
      for a module object: the module's attributes.
      for a class object:  its attributes, and recursively the attributes
        of its bases.
      for any other object: its attributes, its class's attributes, and
        recursively the attributes of its class's base classes.



## 常用的python指令


##### 基础语句

| 指令 | 说明 | 示例 |
|------|------|------|
| `=` | 赋值语句 | `x = 10` |
| `del` | 删除变量或对象元素 | `del lst[0]` |
| `pass` | 空操作占位符 | `def foo(): pass` |
| `assert` | 断言条件为真 | `assert x > 0` |
| `raise` | 抛出异常 | `raise ValueError()` |

##### 流程控制

| 指令 | 说明 | 示例 |
|------|------|------|
| `if/elif/else` | 条件判断 | `if x>0: print("正数")` |
| `for` | 迭代循环 | `for i in range(5):` |
| `while` | 条件循环 | `while x < 10:` |
| `break` | 跳出循环 | `while True: break` |
| `continue` | 跳过当前迭代 | `for i in range(5): if i==2: continue` |
| `try/except/finally` | 异常处理 | `try: x=1/0 except: print("错误")` |

##### 函数与类

| 指令 | 说明 | 示例 |
|------|------|------|
| `def` | 定义函数 | `def foo(): return 1` |
| `return` | 函数返回值 | `return x + y` |
| `yield` | 生成器返回值 | `yield i**2` |
| `class` | 定义类 | `class Foo:` |
| `lambda` | 匿名函数 | `f = lambda x: x+1` |

##### 导入与作用域

| 指令 | 说明 | 示例 |
|------|------|------|
| `import` | 导入模块 | `import os` |
| `from...import` | 从模块导入 | `from math import sqrt` |
| `global` | 声明全局变量 | `global count` |
| `nonlocal` | 声明非局部变量 | `nonlocal x` |

##### 上下文管理

| 指令 | 说明 | 示例 |
|------|------|------|
| `with` | 上下文管理器 | `with open('a.txt') as f:` |
| `as` | 别名赋值 | `import numpy as np` |

##### 其他特殊语句

| 指令 | 说明 | 示例 |
|------|------|------|
| `match/case` (Python 3.10+) | 模式匹配 | `match x: case 1: print("one")` |
| `...` (Ellipsis) | 省略号对象 | `def foo(): ...` |
| `@` | 装饰器语法 | `@decorator def foo(): pass` |


### 赋值语句

- 赋值语句将对象与变量名绑定；
- 变量名存储的不是对象的值，而是对象的内存地址，即变量是对 对象 的引用(reference)，类似于C语言中的指针；

In [158]:
# basic assignment
a = [1, 2, 3]
a

[1, 2, 3]

In [159]:
# sequence assignment
a, b, c = [1, 2, 3]
print(f'a = {a}, b = {b}, c = {c}')

a = 1, b = 2, c = 3


In [160]:
[a, b, c] = [1, 2, 3]
print(f'a = {a}, b = {b}, c = {c}')

a = 1, b = 2, c = 3


In [173]:
a, b, c = range(3)
print(f'a = {a}, b = {b}, c = {c}')

a = 0, b = 1, c = 2


In [174]:
a, b, c = range(2)
print(f'a = {a}, b = {b}, c = {c}')

ValueError: not enough values to unpack (expected 3, got 2)

In [161]:
# extended-unpacking assignment
a, b = [1, 2, 3]

ValueError: too many values to unpack (expected 2)

In [162]:
a, *b = [1, 2, 3]
a, b

(1, [2, 3])

In [163]:
a, *b, c = [1, 2, 3, 4]
a, b, c

(1, [2, 3], 4)

In [179]:
a, b, *c = range(2)
a, b, c

(0, 1, [])

In [178]:
a, b, *c, *d = range(2)
a, b, c

SyntaxError: multiple starred expressions in assignment (991317121.py, line 1)

In [166]:
a = b = c = d = 5
a += 1   # a = a + 1
b *= 2   # b = b * 2
c /= 2   # c = c / 3
d %= 3   # d = d % 3
a, b, c, d

(6, 10, 2.5, 2)

In [167]:
abc

NameError: name 'abc' is not defined

In [171]:
# assignment in expression

b = (a := 3) + 4  # equal to: a = 3; b = a + 3
a, b

(3, 7)

##### 🙋练习

在执行之前你能猜出下面这两段代码的输出是什么吗？
```python
a, *b = 'python'
print([a, b])
```

```python
a = []
a += 'python'
print(a)
```

##### 共享的引用

In [17]:
a = [1, 2, 3]
b = a
print(b)
a[0] = -1
print(b)

[1, 2, 3]
[-1, 2, 3]


In [15]:
a = [1, 2, 3]
b = [a, 4, 5, 6]
print(b)

[[1, 2, 3], 4, 5, 6]


In [16]:
a[0] = -1
print(b)

[[-1, 2, 3], 4, 5, 6]


In [19]:
# 怎么避免
a = [1, 2, 3]
b = [a[:], 4, 5, 6]
print(b)
a[0] = -1
print(b)

[[1, 2, 3], 4, 5, 6]
[[1, 2, 3], 4, 5, 6]


In [21]:
a = [1, 2, 3]
b = a.copy()
print(b)
a[0] = -1
print(b)

[1, 2, 3]
[1, 2, 3]


### print

In [180]:
print('a', 'b')

a b


In [181]:
print('a', 'b', sep=',')

a,b


In [188]:
print('a', 'b', 'c', sep='\t')
print('-----')

a	b	c
-----


In [189]:
print('a', 'b', 'c', sep='\t', end=';')
print('-----')

a	b	c;-----


### 流程控制

开始之前：
- python会自动检测指令或代码块结束的位置；
- 一行的末尾表示一条指令的结束，不需要用特殊的符号比如`;`来结束一条指令；
- 可以用`;`把几个简单指令组合写在一行；
- 如果指令延续多行，可以用括号(`()`, `[]`, `{}`)括起来；
- 在一行的末尾加一个反斜杠也可以表示指令还没结束，需要延续到下一行，但容易出错；(不推荐使用)
- 缩进表示一个代码块(code block)的开始，缩进结束即block的结束；（注意: 在正式的文本中不要中英混杂）；
- 缩进的形式可以用tab、四个空格、两个空格、或tab，但在程序中要保持一致，不要混合使用；

In [114]:
x = [1, 2,
     3, 4]
x

[1, 2, 3, 4]

In [116]:
y = 'a' 'b'
'c'
print(y)

ab


In [117]:
y = ('a' 'b'
     'c')
print(y)

abc


#### if-else

```python
if test:
    code block
elif test:
    code block
else:
    code block
```

In [106]:
x = 3
if x > 0:
    print("x is positive!")

x is positive!


In [109]:
x = -2
if x > 0:
    print("x is positive!")
else:
    print("x is not positive!")

x is not positive!


In [110]:
x = -2
if x > 0:
    print("x is positive!")
elif x == 0:
    print("x is zero!")
else:
    print("x is negative!")

x is negative!


In [113]:
x = -2
if x > 0:
    print("x is positive!")
elif x == 0:
    print("x is zero!")
else:
    print("x is negative!")

x is negative!


In [193]:
# multiple choices
a = 2
match a:
    case 1:
        print('first option')
    case 2:
        print('second option')
    case 3:
        print('third option')

second option


In [2]:
a = 4
match a:
    case 1:
        print('first option')
    case 2:
        print('second option')
    case 3:
        print('third option')
    case _:
        print('invalid option')
        # sys.exit(1)

invalid option


In [4]:
score_a, score_b, score_c = 100, 80, 61
if score_a > 60 and score_b > 60 and score_c > 60:
    print('pass')
else:
    print('fail')

pass


In [3]:
# 大小比较返回True，False (即返回整数1或0）
2 < 3, 2 > 3

(True, False)

In [6]:
# `and` 返回什么？
a = score_a > 60 and score_b > 60

In [8]:
a = 2 and 3
print(a)

3


In [9]:
a = 3 and 2
print(a)

2


##### 反思and、or、not

- python中的任何对象都有一个内在的`True`或`False`值；
- 数字0、空对象、`None`的布尔值是`False`，其他对象的布尔值是`True`；
- 比较(>, <, >=, <=)或相等（==）测试返回的值是`True`/`False`；
- `or` : 如果左侧对象的布尔值是`True`，返回左边的对象，否则返回右边的对象；
- `and`: 如果左侧对象的布尔值是`False`，返回左边的对象，否则返回右边的对象；
- `not`: 返回与右侧对象的布尔值相反的布尔值；

🙋**练习**

判断下面代码的数据结果
```python
a = 0 or 1
print(a)
```

##### is和==

- `==` 比较两个对象的值是否相等；
- `is` 比较两个对象是否相同（位于相同的内存地址）；

In [13]:
a = b = 'Hello, World!'
c = 'Hello, World!'
a is b, a is c

(True, False)

In [24]:
# 查看a,b,c的内存地址
id(a), id(b), id(c)

(4384459584, 4384075584, 4384459696)

In [14]:
a == b, a == c

(True, True)

🙋**练习**

判断下面代码的数据结果是True还是False？
```python
a = 'py'
b = 'py'
print(a is b)
```
一定要做一下，没有看起来那么简单！想想为什么会输出这样的结果？(提示：用`id`检验自己的猜想)

In [25]:
# 类似于C语言中的条件赋值如何实现？
#   condition? a : b
a = 2
b = 'positive' if a > 0 else 'non-positive'
print(b)

positive


#### While

```python
while test:
    ...
    if test: break
    if test: continue
else:
    ...  # 循环正常结束时(即不通过break结束）执行
```

注意：**不要写死循环！**

In [ ]:
# do not run this cell in notebook
while True:
    x = input("Please enter a number:")
    if x == "q": break
    print(float(x)**2)
print("Bye!")

In [28]:
y = []
x = 1
while x < 10:
    print(f'x = {x}', end=', ')
    y.append(x**2)
    x += 1

y

x = 1, x = 2, x = 3, x = 4, x = 5, x = 6, x = 7, x = 8, x = 9, 

[1, 4, 9, 16, 25, 36, 49, 64, 81]

In [153]:
from time import sleep

y = []
x = 1

while x < 10:
    print(f'x = {x}', end=', ')
    y.append(x**2)
    # x += 1
    if x >= 10: break  # 
    sleep(1)

y

x = 1
x = 1
x = 1
x = 1


KeyboardInterrupt: 

#### for-loop

```python
for item in iterable:
    ...
    if test: break
    if test: continue
else:
    ...
```

可迭代对象(iterable)是可以遍历的对象，本质上来说是实现了`__iter__`方法的对象，包括各种序列(list, tuple, range)以及dict等以及迭代器。

In [39]:
x = [1, 'a', 'abc']
'__iter__' in dir(x)

True

In [40]:
y = range(10)
'__iter__' in dir(y)

True

In [33]:
for i in x:
    print(i)

1
a
abc


In [128]:
x = dict(zip('abc', ['1', '2', '3']))
x

{'a': '1', 'b': '2', 'c': '3'}

In [132]:
for i in x:
    print(i)

a
b
c


In [130]:
x.keys()

dict_keys(['a', 'b', 'c'])

In [152]:
x = []
for i in range(10):
    print(f'i = {i}')
    if i % 2 == 1:
        continue   # 如果是奇数，当前循环结束，继续下一个循环
    x.append(i**2)

x

i = 0
i = 1
i = 2
i = 3
i = 4
i = 5
i = 6
i = 7
i = 8
i = 9


[0, 4, 16, 36, 64]

In [42]:
d = {'a': 1, 2: 'b'}
for key, value in d.items():
    print(f'{key} => {value}')

a => 1
2 => b


##### for-loop相关的辅助函数

```
range(from, to, step_size=1)

zip(iterable, iterable, ...)

enumerate(iterable)
```

In [44]:
range(10)

range(0, 10)

In [43]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [45]:
list(range(1, 10))

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [46]:
list(range(1, 10, 2))

[1, 3, 5, 7, 9]

In [52]:
x = range(5)
y = 'abcde'
zip(x, y)

In [53]:
list(zip(x, y))

[(0, 'a'), (1, 'b'), (2, 'c'), (3, 'd'), (4, 'e')]

In [56]:
# length is determined by the shortest iterable
list(zip(range(3), 'abcde'))

[(0, 'a'), (1, 'b'), (2, 'c')]

In [50]:
z = list('ABCDE')

for i in zip(x, y, z):
    print(i)

(0, 'a', 'A')
(1, 'b', 'B')
(2, 'c', 'C')
(3, 'd', 'D')
(4, 'e', 'E')


In [51]:
for a, *b in zip(x, y, z):
    print(a, b)

0 ['a', 'A']
1 ['b', 'B']
2 ['c', 'C']
3 ['d', 'D']
4 ['e', 'E']


In [57]:
index = 0
for char in 'abcde':
    print(f'character at index {index} is {char}')
    index += 1

character at index 0 is a
character at index 1 is b
character at index 2 is c
character at index 3 is d
character at index 4 is e


In [58]:
for i in enumerate('abcde'):
    print(i)

(0, 'a')
(1, 'b')
(2, 'c')
(3, 'd')
(4, 'e')


In [59]:
for index, char in enumerate('abcde'):
    print(f'character at index {index} is {char}')

character at index 0 is a
character at index 1 is b
character at index 2 is c
character at index 3 is d
character at index 4 is e


#### try-except
捕获可能出现的代码错误，并采取必要的措施，防止程序中断；

In [135]:
print(1/0)

ZeroDivisionError: division by zero

In [136]:
try:
    print(1/0)
except:
    print('Division error')

Division error


In [137]:
x = 1
y = 0
try:
    z = 1/0
except ZeroDivisionError:
    print('division by zero')
    z = 100

division by zero


In [138]:
print(z)

100


In [139]:
d = dict()
try:
    x = d['A']
except KeyError:
    print('"A" is not a valid key in d')

"A" is not a valid key in d


## 可迭代对象（iterable）

- 可迭代对象（iterable）: 有`__iter__`方法的对象，可以遍历生成序列；
- 迭代器（iterator）: 有`__next__`方法的对象，每次运行都会给出对象下一次迭代的值；


In [80]:
x = enumerate(range(1, 5))
', '.join(dir(x))

'__class__, __class_getitem__, __delattr__, __dir__, __doc__, __eq__, __format__, __ge__, __getattribute__, __getstate__, __gt__, __hash__, __init__, __init_subclass__, __iter__, __le__, __lt__, __ne__, __new__, __next__, __reduce__, __reduce_ex__, __repr__, __setattr__, __sizeof__, __str__, __subclasshook__'

In [81]:
list(x)

[(0, 1), (1, 2), (2, 3), (3, 4)]

In [82]:
x = enumerate(range(1, 5))
a = list(x)
b = list(x)
a == b

False

In [83]:
next(x)

StopIteration: 

In [84]:
x = enumerate(range(1, 5))

In [85]:
next(x)

(0, 1)

In [86]:
next(x)

(1, 2)

In [87]:
list(x)  # 迭代器中已经遍历过的值没了！

[(2, 3), (3, 4)]

In [95]:
list(x)  # # 迭代器会耗竭

['a', 'b', 'c', 'd', 'e']

In [89]:
x = 'abcde'
y = iter(x)  # 把iterable转为iterator

In [90]:
next(x)

TypeError: 'str' object is not an iterator

In [91]:
next(y)

'a'

In [92]:
next(y)

'b'

In [93]:
list(x), list(y)

(['a', 'b', 'c', 'd', 'e'], ['c', 'd', 'e'])

In [145]:
x = 'aaabbbcccdddeeefff'
list(map(''.join, zip(*[iter(x)]*3)))
# 练习题：分解代码，逐步执行，看看怎样一步步得到这个结果的

['aaa', 'bbb', 'ccc', 'ddd', 'eee', 'fff']

##### 函数迭代(functional itrable)

In [98]:
x = map(int, '12345')

In [99]:
', '.join(dir(x))

'__class__, __delattr__, __dir__, __doc__, __eq__, __format__, __ge__, __getattribute__, __getstate__, __gt__, __hash__, __init__, __init_subclass__, __iter__, __le__, __lt__, __ne__, __new__, __next__, __reduce__, __reduce_ex__, __repr__, __setattr__, __sizeof__, __str__, __subclasshook__'

In [100]:
print(next(x))

1


In [101]:
list(x)

[2, 3, 4, 5]

## Python 列表推导式（List Comprehension）

In [104]:
y = []
for i in range(10):
    y.append(i**2)
y

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [105]:
[i**2 for i in range(10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [128]:
%%timeit -n 10 -r 5
# -n: 每轮重复执行多少次代码；
# -r: 一共执行几轮；
y = []
for i in range(100_0000):
    y.append(i**2)

44.6 ms ± 1.06 ms per loop (mean ± std. dev. of 5 runs, 10 loops each)


In [129]:
%timeit -n 10 -r 5 [i**2 for i in range(100_0000)]

40.2 ms ± 307 μs per loop (mean ± std. dev. of 5 runs, 10 loops each)


In [130]:
y = []
for i in range(10):
    if i % 2 == 0:
        y.append(i**2)
y

[0, 4, 16, 36, 64]

In [132]:
# with if
[i**2 for i in range(10) if i % 2 == 0]

[0, 4, 16, 36, 64]

In [135]:
# nested for loop
x = [i + j + k for i in 'ACGT' for j in 'ACGT' for k in 'ACGT']
print(x)

['AAA', 'AAC', 'AAG', 'AAT', 'ACA', 'ACC', 'ACG', 'ACT', 'AGA', 'AGC', 'AGG', 'AGT', 'ATA', 'ATC', 'ATG', 'ATT', 'CAA', 'CAC', 'CAG', 'CAT', 'CCA', 'CCC', 'CCG', 'CCT', 'CGA', 'CGC', 'CGG', 'CGT', 'CTA', 'CTC', 'CTG', 'CTT', 'GAA', 'GAC', 'GAG', 'GAT', 'GCA', 'GCC', 'GCG', 'GCT', 'GGA', 'GGC', 'GGG', 'GGT', 'GTA', 'GTC', 'GTG', 'GTT', 'TAA', 'TAC', 'TAG', 'TAT', 'TCA', 'TCC', 'TCG', 'TCT', 'TGA', 'TGC', 'TGG', 'TGT', 'TTA', 'TTC', 'TTG', 'TTT']


In [136]:
print([i for i in x if not i in ['TAA', 'TAG', 'TGA']])

['AAA', 'AAC', 'AAG', 'AAT', 'ACA', 'ACC', 'ACG', 'ACT', 'AGA', 'AGC', 'AGG', 'AGT', 'ATA', 'ATC', 'ATG', 'ATT', 'CAA', 'CAC', 'CAG', 'CAT', 'CCA', 'CCC', 'CCG', 'CCT', 'CGA', 'CGC', 'CGG', 'CGT', 'CTA', 'CTC', 'CTG', 'CTT', 'GAA', 'GAC', 'GAG', 'GAT', 'GCA', 'GCC', 'GCG', 'GCT', 'GGA', 'GGC', 'GGG', 'GGT', 'GTA', 'GTC', 'GTG', 'GTT', 'TAC', 'TAT', 'TCA', 'TCC', 'TCG', 'TCT', 'TGC', 'TGG', 'TGT', 'TTA', 'TTC', 'TTG', 'TTT']


In [138]:
# 结果可以是其他类型，不限于list
{i**2 for i in range(10)}

{0, 1, 4, 9, 16, 25, 36, 49, 64, 81}

In [139]:
{i:i**2 for i in range(10)}

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}

In [ ]:
(i**2 for i in range(10))  # generator, not tuple

<generator object <genexpr> at 0x108ad4110>

## 扩展 （自学内容）

### 基础对象的拓展

- https://docs.python.org/3/library/collections.html


###  为高效循环创建迭代器的函数

- https://docs.python.org/3/library/itertools.html (english)
- https://docs.python.org/zh-cn/3.12/library/itertools.html (中文）